jax/_src/debugging.py

```diff
@@ -0,0 +1,147 @@
+batching.primitive_batchers[debug_callback_p] = debug_callback_batching_rule
+
+def debug_callback_jvp_rule(*flat_args, callback: Callable[..., Any],
+    effect: DebugEffect, in_tree: tree_util.PyTreeDef, out_avals: Any):
+  del flat_args, callback, effect, in_tree, out_avals
+  # TODO(sharadmv): implement jvp rule
+  raise NotImplementedError('JVP not supported for `debug_callback`.')
+ad.primitive_jvps[debug_callback_p] = debug_callback_jvp_rule
+
+def debug_callback_transpose_rule(*flat_args, callback: Callable[..., Any],
+    effect: DebugEffect, in_tree: tree_util.PyTreeDef, out_avals: Any):
+  del flat_args, callback, effect, in_tree, out_avals
+  # TODO(sharadmv): implement transpose rule
+  raise NotImplementedError('Transpose not supported for `debug_callback`.')
+ad.primitive_transposes[debug_callback_p] = debug_callback_transpose_rule
+
+# TODO(sharadmv): remove this global keepalive list in favor of attaching
+# keepalives to the module context.
+_keepalives = []
+
+def _ordered_effect_lowering(ctx, token, *args, **params):
+  avals_in = [core.abstract_token, *ctx.avals_in]
+  avals_out = [core.abstract_token, *ctx.avals_out]
+  args = (token, *args)
+  def _callback(token, *flat_args):
+    out = debug_callback_p.impl(*flat_args, **params)
+    return (token, *out)
+  (token, *result), keepalive = mlir.emit_python_callback(
+      ctx.module_context.platform, _callback, list(args), avals_in, avals_out,
+      True)
+  return result, keepalive, token
+
+def debug_callback_lowering(ctx, *args, effect, callback, **params):
+  if effect in core.ordered_effects:
+    token = ctx.tokens_in.get(effect)[0]
+    result, keepalive, token = _ordered_effect_lowering(ctx, token,
+        *args, effect=effect, callback=callback, **params)
+    ctx.set_tokens_out(mlir.TokenSet({effect: (token,)}))
+  else:
+    def _callback(*flat_args):
+      return tuple(debug_callback_p.impl(
+        *flat_args, effect=effect, callback=callback, **params))
+    result, keepalive = mlir.emit_python_callback(ctx.module_context.platform,
+      _callback, list(args), ctx.avals_in, ctx.avals_out,  True)
+  _keepalives.append(keepalive)
+  return result
+mlir.register_lowering(debug_callback_p, debug_callback_lowering,
+                       platform="cpu")
+
+def debug_callback(callback: Callable[..., Any],
+                   effect: DebugEffect, result_shapes: Any, *args, **kwargs):
+  """Calls a stageable Python callback.
+
+  `debug_callback` enables you to pass in a Python function that can be called
+  inside of a staged JAX program.
+
+  Args:
+    callback: A Python callable.
+    effect: A `DebugEffect`.
+    result_shapes: A Pytree of `jax.ShapeDtypeStruct`-like objects that match
+      output of `callback`.
+      *args: The positional arguments to the callback.
+      **kwargs: The positional arguments to the callback.
+  Returns:
+    The results of `callback` applied to `*args` and `**kwargs`.
```

Comment:
```suggestion
    The value of `callback(*args, **kwargs)`.
```

What are the use cases for a debug callback returning anything though?

Reply:
None for now, but the callback API does support it.